proof/invariant-abstract/ARM_HYP/ArchArch_AI.thy

```diff
@@ -983,24 +987,196 @@ lemma valid_global_vspace_mappings_vcpu_update_str:
   "valid_global_vspace_mappings s \<Longrightarrow> valid_global_vspace_mappings (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>)"
   by (simp add: valid_global_vspace_mappings_def)
 
+lemma arm_current_vcpu_update_valid_global_vspace_mappings[simp]:
+  "valid_global_vspace_mappings (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>)
+   = valid_global_vspace_mappings s"
+  by (clarsimp simp: valid_global_vspace_mappings_def global_refs_def
+              split: kernel_object.splits option.splits)
+
+crunches associate_vcpu_tcb
+  for pspace_aligned[wp]: pspace_aligned
+  and pspace_distinct[wp]: pspace_distinct
+  and zombies_final[wp]: zombies_final
+  and sym_refs_state_refs_of[wp]: "\<lambda>s. sym_refs (state_refs_of s)"
+  and valid_mdb[wp]: valid_mdb
+  and valid_ioc[wp]: valid_ioc
+  and only_idle[wp]: only_idle
+  and if_unsafe_then_cap[wp]: if_unsafe_then_cap
+  and valid_reply_caps[wp]: valid_reply_caps
+  and valid_reply_masters[wp]: valid_reply_masters
+  and valid_irq_node[wp]: valid_irq_node
+  and valid_irq_handlers[wp]: valid_irq_handlers
+  and cap_refs_in_kernel_window[wp]: cap_refs_in_kernel_window
+  and pspace_respects_device_region[wp]: pspace_respects_device_region
+  and cur_tcb[wp]: cur_tcb
+  and valid_vspace_objs[wp]: valid_vspace_objs
+  and valid_arch_caps[wp]: valid_arch_caps
+  and valid_kernel_mappings[wp]: valid_kernel_mappings
+  and equal_kernel_mappings[wp]: equal_kernel_mappings
+  and valid_asid_map[wp]: valid_asid_map
+  and valid_global_vspace_mappings[wp]: valid_global_vspace_mappings
+  and pspace_in_kernel_window[wp]: pspace_in_kernel_window
+  (wp: crunch_wps dmo_valid_irq_states device_region_dmos simp: crunch_simps)
+
+crunches vcpu_switch
+  for valid_idle[wp]: valid_idle
+  (wp: crunch_wps)
+
+lemma associate_vcpu_tcb_valid_idle[wp]:
+  "\<lbrace>valid_idle and (\<lambda>s. tcb \<noteq> idle_thread s)\<rbrace> associate_vcpu_tcb vcpu tcb \<lbrace>\<lambda>_. valid_idle\<rbrace>"
+  apply (clarsimp simp: associate_vcpu_tcb_def)
+  by (wp hoare_vcg_all_lift
+      | wp (once) hoare_drop_imps
+      | wpc
+      | simp add: dissociate_vcpu_tcb_def vcpu_invalidate_active_def)+
+
+lemma arm_current_vcpu_update_valid_global_refs[simp]:
+  "valid_global_refs (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>) = valid_global_refs s"
+  by (clarsimp simp: valid_global_refs_def global_refs_def
+              split: kernel_object.splits option.splits)
+
+crunches associate_vcpu_tcb
+  for valid_global_refs[wp]: valid_global_refs
+  (wp: crunch_wps)
+
+crunches vcpu_restore_reg
+  for valid_irq_states[wp]: valid_irq_states
+  (wp: crunch_wps dmo_valid_irq_states simp: writeVCPUHardwareReg_def)
+
+method repeat_unless methods stop cont
+  = (stop | (cont, repeat_unless stop cont))
+
```

Comment:
Cool, nice one. This one looks more generally useful and could move up to somewhere in Lib where we define more Eisbach methods (maybe with a small example).

Reply:
Maybe swap the arguments? With the name `repeat_unless` I would expect the `cont` first and the `stop` second.

Reply:
This (as well as `hoare_seq_ext_skip`) is just moved from Lib. They mustn't have found their way into `master` yet. There aren't many uses of `repeat_unless` in `rt`. I agree that the arguments would be better swapped, but should we do it both here and on `rt`, or leave it for now?

Reply:
Hm, that changes things. Ideally we'd change it in `rt` first and then backport the Lib change, but that sounds like a bunch of work that is a distraction.

Let's maybe at least find the `rt` commit that puts it into Lib and cherry-pick that onto this branch, so that it ends up in master when we merge here. This should avoid a merge conflict for updates later. For the argument order, let's leave it as is for now, maybe with a FIXME. It is documented, just not the order of least surprise.

Reply:
Probably worth to point out that with this argument order and definition this looks more like `repeat_until`. It repeats `cont` until `stop` applies. So I would second putting in a FIXME and reconsidering merge with RT later.